Tools/LocomotionTool/LocomotionTool.cs

```diff
@@ -0,0 +1,599 @@
+					return;
+			}
+
+			if (DoTwoHandedScaling(consumeControl))
+				return;
+
+			if (m_Preferences.blinkMode)
+			{
+				if (DoBlink(consumeControl))
+					return;
+			}
+			else
+			{
+				if (DoFlying(consumeControl))
+					return;
+			}
+
+			DoCrawl(consumeControl);
+		}
+
+		bool DoFlying(ConsumeControlDelegate consumeControl)
+		{
+			var reverse = m_LocomotionInput.reverse.isHeld;
+			var moving = m_LocomotionInput.forward.isHeld || reverse;
+			if (moving)
+			{
+				if (m_LocomotionInput.rotate.isHeld)
+				{
+					var localRayRotation = Quaternion.Inverse(cameraRig.rotation) * rayOrigin.rotation;
+					var localRayForward = localRayRotation * Vector3.forward;
+					if (Mathf.Abs(Vector3.Dot(localRayForward, Vector3.up)) > k_RotationThreshold)
+						return true;
+
+					localRayForward.y = 0;
+					localRayForward.Normalize();
+					if (!m_Rotating)
+					{
+						m_Rotating = true;
+						m_RigStartPosition = cameraRig.position;
+						m_RigStartRotation = cameraRig.rotation;
+
+						m_RayOriginStartForward = localRayForward;
+						m_RayOriginStartRight = localRayRotation * (reverse ? Vector3.left : Vector3.right);
+						m_RayOriginStartRight.y = 0;
+						m_RayOriginStartRight.Normalize();
+
+						m_CameraStartPosition = CameraUtils.GetMainCamera().transform.position;
+						m_LastRotationDiff = Quaternion.identity;
+					}
+
+					consumeControl(m_LocomotionInput.rotate);
+					var startOffset = m_RigStartPosition - m_CameraStartPosition;
+
+					var angle = Vector3.Angle(m_RayOriginStartForward, localRayForward);
+					var dot = Vector3.Dot(m_RayOriginStartRight, localRayForward);
+					var rotation = Quaternion.Euler(0, angle * Mathf.Sign(dot), 0);
+					var filteredRotation = Quaternion.Lerp(m_LastRotationDiff, rotation, k_RotationDamping);
+
+					const float segmentSize = 360f / k_RotationSegments;
+					var segmentedRotation = Quaternion.Euler(0, Mathf.Round(filteredRotation.eulerAngles.y / segmentSize) * segmentSize, 0);
+
+					cameraRig.rotation = m_RigStartRotation * segmentedRotation;
+					cameraRig.position = m_CameraStartPosition + segmentedRotation * startOffset;
+
+					m_LastRotationDiff = filteredRotation;
+				}
+				else
+				{
+					var speed = k_SlowMoveSpeed;
+					var speedControl = m_LocomotionInput.speed;
+					var speedControlValue = speedControl.value;
+					if (!Mathf.Approximately(speedControlValue, 0)) // Consume control to block selection
+					{
+						speed = k_SlowMoveSpeed + speedControlValue * (k_FastMoveSpeed - k_SlowMoveSpeed);
+						consumeControl(speedControl);
+					}
+
+					speed *= this.GetViewerScale();
+					if (reverse)
+						speed *= -1;
+
+					m_Rotating = false;
+					cameraRig.Translate(Quaternion.Inverse(cameraRig.rotation) * rayOrigin.forward * speed * Time.unscaledDeltaTime);
+				}
+
+				consumeControl(m_LocomotionInput.forward);
+				return true;
+			}
+
+			m_Rotating = false;
+			return false;
+		}
+
+		bool DoCrawl(ConsumeControlDelegate consumeControl)
```

Comment:
Doesn't look like we're using this return value.  Though it does match the "Do" functions.